pkg/operation/seed/seed.go

```diff
@@ -1853,3 +1869,95 @@ func cleanupOrphanExposureClassHandlerResources(ctx context.Context, c client.Cl
+		return false, client.IgnoreNotFound(err)
+	}
+
+	switch newStorageQuantity.Cmp(*pvc.Spec.Resources.Requests.Storage()) {
+	case 0:
+		return false, nil
+	case 1:
+		// json merge patch
+		patch := client.MergeFrom(pvc.DeepCopy())
+		pvc.Spec.Resources.Requests = corev1.ResourceList{
+			corev1.ResourceStorage: newStorageQuantity,
+		}
+		log.Infof("Patch garden/loki PVC %q to %q of storage", pvc.Name, newStorageQuantity.String())
+		if err := k8sClient.Patch(ctx, pvc, patch); err != nil {
+			return false, err
+		}
+		return true, nil
+	case -1:
+		log.Infof("Delete garden/loki PVC %q", pvc.Name)
+		if err := k8sClient.Delete(ctx, pvc); err != nil {
+			return false, client.IgnoreNotFound(err)
+		}
+		return true, nil
+	default:
+		// This should never happen
+		return false, fmt.Errorf("unknow result comparing PVC %s storage value %v with %v", pvcName, pvc.Spec.Resources.Requests.Storage(), newStorageQuantity)
+	}
+}
+
+func prepareLokiPVCForResizing(ctx context.Context, k8sClient client.Client, newStorageQuantity resource.Quantity, log logrus.FieldLogger) error {
+	pvc := &corev1.PersistentVolumeClaim{}
+	if err := k8sClient.Get(ctx, kutil.Key(v1beta1constants.GardenNamespace, "loki-loki-0"), pvc); err != nil {
+		return client.IgnoreNotFound(err)
+	}
+
+	if newStorageQuantity.Cmp(*pvc.Spec.Resources.Requests.Storage()) == 0 {
+		return nil
+	}
+
+	log.Infof("Scaling StatefulState %s/%s to zero in order to detach pvc %q", v1beta1constants.GardenNamespace, v1beta1constants.StatefulSetNameLoki, pvc.Name)
+	if err := kubernetes.ScaleStatefulSetAndWaitUntilIsScaled(ctx, k8sClient, kutil.Key(v1beta1constants.GardenNamespace, v1beta1constants.StatefulSetNameLoki), 0); client.IgnoreNotFound(err) != nil {
+		return err
+	}
+
+	volumeAttachment, err := getVolumeAttachmetByPVC(ctx, k8sClient, pvc)
+	if err != nil {
+		return err
+	}
+
+	if volumeAttachment == nil {
+		return nil
+	}
+
+	log.Infof("Waiting VolumeAttachment %q to be deleted", volumeAttachment.Name)
+	return waitUntilVolumeAttachmentIsDeleted(ctx, k8sClient, kutil.Key(volumeAttachment.Name))
+}
+
+func getVolumeAttachmetByPVC(ctx context.Context, k8sClient client.Client, pvc *corev1.PersistentVolumeClaim) (*storagev1.VolumeAttachment, error) {
+	volumeattachments := &storagev1.VolumeAttachmentList{}
+	if err := k8sClient.List(ctx, volumeattachments); err != nil {
+		return nil, err
+	}
+
+	for _, volumeattachment := range volumeattachments.Items {
+		if volumeattachment.Spec.Source.PersistentVolumeName != nil &&
+			*volumeattachment.Spec.Source.PersistentVolumeName == pvc.Spec.VolumeName {
+			return &volumeattachment, nil
+		}
+	}
+
+	return nil, nil
+}
+
+func waitUntilVolumeAttachmentIsDeleted(ctx context.Context, c client.Client, key types.NamespacedName) error {
+	return retry.UntilTimeout(ctx, 5*time.Second, 300*time.Second, func(ctx context.Context) (done bool, err error) {
+		volumeattachment := &storagev1.VolumeAttachment{}
```

Comment:
Does this resource always exist even if the seed cluster is not using CSI?

Reply:
I made one shoot on azure version 1.17 and the `volumeattachment` resource type was there.
If the object of that resource is not present it will return success, but the behavior will be unpredicted.
I have no other idea how to know if the PV is detached or not.

Reply:
The resource definition exists, but there are no objects. So I think it cannot be used to check whether the volume got detached for clusters without CSI. Maybe you can use the `.status.volumesAttached` field of the respective `Node` object to get the information you need?

Reply:
I have made a new commit where I am using the `.status.volumesAttached` but it didn't work as I expected for Azure.
When I tried to increase the size from 60Gi up to 140Gi with 10Gi steps after 90Gi and 100Gi I got an error that the volume was already attached and can't be expanded or there is no external plugin to handle the request but after 5 to 10 minutes the PV was resized and then the PVC. After 110G I was not able to increase the storage:
```text
Warning  VolumeResizeFailed          33m (x23 over 67m)    external-resizer disk.csi.azure.com  resize volume "pv-shoot--i330716--azure-seed-2d32785a-ebba-43ed-bf13-d92c15e94555" by resizer "disk.csi.azure.com" failed: rpc error: code = Internal desc = failed to resize disk(/subscriptions/00d2caa5-cd29-46f7-845a-2f8ee0360ef5/resourceGroups/shoot--i330716--azure-seed/providers/Microsoft.Compute/disks/pv-shoot--i330716--azure-seed-2d32785a-ebba-43ed-bf13-d92c15e94555) with error(azureDisk - disk resize is only supported on Unattached disk, current disk state: Attached, already attached to /subscriptions/00d2caa5-cd29-46f7-845a-2f8ee0360ef5/resourceGroups/shoot--i330716--azure-seed/providers/Microsoft.Compute/virtualMachines/shoot--i330716--azure-seed-worker-fzi10-z2-fbb8c-wnsz2)
  Normal   FileSystemResizeSuccessful  25m (x4 over 71m)     kubelet                              MountVolume.NodeExpandVolume succeeded for volume "pv-shoot--i330716--azure-seed-2d32785a-ebba-43ed-bf13-d92c15e94555"
  Normal   Resizing                    6m49s (x58 over 81m)  external-resizer disk.csi.azure.com  External resizer is resizing volume pv-shoot--i330716--azure-seed-2d32785a-ebba-43ed-bf13-d92c15e94555
  Warning  ExternalExpanding           3m31s (x8 over 81m)   volume_expand                        Ignoring the PVC: didn't find a plugin capable of expanding the volume; waiting for an external controller to process this PVC.
``` 

It looks like because of the Azure we are not able to use the disk expansion as it is documented.
The logs are not so important and the increase of the volume size will be very rare, so I am thinking can we simply delete the disk every time we want to increase it?

Reply:
cc @dguendisch @istvanballok opinions?

Reply:
> The logs are not so important and the increase of the volume size will be very rare, so I am thinking can we simply delete the disk every time we want to increase it?

Last time we threw away the log disks (when we enabled disk encryption) we notified everyone in advance. I'd hope that throwing away disks is however a very rare operation and only done for monitoring/logs when the dev efforts would be unreasonably high to implement resizing.

Isn't your test showing that resizing "basically" works on azure but fails with many consecutive tries? Wouldn't it make sense to do the resize in one step, and if that fails for a **few** volumes we either notice and manually clean it up or we don't notice in which case nothing bad has happened other then the volume not being expanded?

Reply:
Initially, the resizing central Loki PR(which was reverted) didn't use any complicated logic, because according to the k8s documentation the pod can be running during expansion. But this is not true for Azure where the PV first needs to be detached. I have tried to simply scale down the Loki to zero in order to detach the PV, but the `gardenlet` is so fast that the new Loki statefulset is deployed before the PV has been fully detached from the node, and resizing fails from time to time mostly on Azure. So to be 100% sure that the resizing is done just after the PV is detached I have watched the `VollumeAttachmet` resource. This gave 100% success, but this resource is used only when we are using CSI drivers. Then @rfranzke suggested using `Node.status.volumesAttached`, which makes the PR more difficult because now we have to find out the volume ID, which could be found in the PV. The PV supports a lot of volume types so we have to search in all of the volume types for the ID. For SCI volume the ID is named volumeHandle, in AzureDisk is DiskName. Who knows what will be the volume ID in the next used type which will be used for `Node.status.volumesAttached`. Next, I have to know to which Node the PV is attached. Should I take it from the pod(what if the pod is scheduled to a node A but the disk is attached to node B) or should I look in all nodes regardless of how much they are and how much volume we have attached to them. Should I look into `Node.status.volumesInUse`, because once I debugged an issue where the volume was not attached but was in use.
During the development, I noticed that the annotation `volume.kubernetes.io/selected-node` of the PVC was showing a non-existing node. And all of this complexity just to expand the central Loki. The end-user logs will not be affected. So only the developer must be warned. So the answer is yes, we can fix it manually but as I said the initial PR was reverted in order to avoid manual fixes.

Reply:
If it gets that complicated I'd personally vote for just resizing the pv, doing nothing more and just live with smaller azure disks for a while. I mean all of our seeds eventually roll their nodes (due to autoscaling, due to OS/K8s updates), so eventually even azure volumes will be detached sooner or later and should then resize ultimately.